### PR TITLE
Feature/enable charge by provider

### PR DESCRIPTION
**Pull request by FredrikAreschoug**

### CODE REVIEW

File: src/main/kotlin/com/hedvig/accountservice/schedulers/web/ScheduleController.kt [hunk 1]
```diff
@@ -199,22 +196,12 @@ class ScheduleController @Autowired constructor(
     return ResponseEntity.noContent().build()
   }
 
-  @PostMapping("/charges")
-  fun chargeAllMembers(
+  @PostMapping("/{paymentProvider}/charges")
```
Comment (meletios) [inline src/main/kotlin/com/hedvig/accountservice/schedulers/web/ScheduleController.kt:199]:
```suggestion
  @PostMapping("paymentProvider/{paymentProvider}/charge")
```

Comment (meletios) [reply src/main/kotlin/com/hedvig/accountservice/schedulers/web/ScheduleController.kt:199]: the name of the resource first / the actual resouce (trustly etc) and then the action

File: src/main/kotlin/com/hedvig/accountservice/schedulers/web/dto/PaymentProvider.kt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.hedvig.accountservice.schedulers.web.dto
+
+enum class PaymentProvider {
+    TRUSTLY,
+    ADYEN,
+    ALL
```
Comment (meletios) [inline src/main/kotlin/com/hedvig/accountservice/schedulers/web/dto/PaymentProvider.kt:6]: another suggestion is to keep /charges and use the enum only for specific payment provider. Or?

File: src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import org.springframework.stereotype.Service
+
+@Service
+class ScheduleChargesServiceImpl(
+    private val commandGateway: CommandGateway,
+    private val stateManager: SchedulerStateManager,
+    private val paymentService: PaymentService
+): ScheduleChargesService {
+
+    override fun scheduleChargesOnPaymentProvider(paymentProvider: PaymentProvider, scheduledBy: String) {
```
Comment (meletios) [inline src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt:21]:
```suggestion
    override fun scheduleChargesByPaymentProvider(paymentProvider: PaymentProvider, scheduledBy: String) {
```

File: src/main/kotlin/com/hedvig/accountservice/integration/paymentService/PaymentServiceClient.kt [hunk 4]
```diff
@@ -18,4 +20,9 @@ interface PaymentServiceClient {
         @PathVariable("memberId") memberId: String,
         @RequestBody chargeRequestDto: ChargeRequestDto
     ): ResponseEntity<UUID>
+
+    @GetMapping("/v2/_/members/connectedToProvider/{paymentProvider}")
+    fun getMembersConnectedToProvider(
```
Comment (vonElfvin) [inline src/main/kotlin/com/hedvig/accountservice/integration/paymentService/PaymentServiceClient.kt:25]: Maybe?
```suggestion
    fun getMembersByProvider(
```

File: src/main/kotlin/com/hedvig/accountservice/integration/paymentService/PaymentServiceClient.kt [hunk 5]
```diff
@@ -18,4 +20,9 @@ interface PaymentServiceClient {
...
+        @PathVariable("paymentProvider") paymentProvider: PaymentProvider
```
Comment (vonElfvin) [inline src/main/kotlin/com/hedvig/accountservice/integration/paymentService/PaymentServiceClient.kt:26]: Super subjective taste question
```suggestion
        @PathVariable("provider") provider: PaymentProvider
```

File: src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+import com.hedvig.accountservice.schedulers.model.SchedulerStateManager
+import com.hedvig.accountservice.schedulers.model.SchedulerStateTransition
+import com.hedvig.accountservice.schedulers.model.SchedulerStateTransitionOutcome
+import com.hedvig.accountservice.schedulers.web.dto.PaymentProvider
+import org.axonframework.commandhandling.gateway.CommandGateway
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Service
+
+@Service
+class ScheduleChargesServiceImpl(
```
Comment (vonElfvin) [inline src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt:15]: Ah yea, back in the day I didn't really know how to structure the code. Logic in the controllers and the sagas 😅

File: src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    private val commandGateway: CommandGateway,
+    private val stateManager: SchedulerStateManager,
+    private val paymentService: PaymentService
+): ScheduleChargesService {
+    override fun scheduleChargesForAll(scheduledBy: String) =
+        scheduleCharges(stateManager.findAllMemberIds(ChargeStatus.APPROVED_FOR_CHARGE), scheduledBy)
```
Comment (vonElfvin) [inline src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt:21]: Some boyscouting maybe

```suggestion
        scheduleCharges(stateManager.findAllMemberIdsByStatus(ChargeStatus.APPROVED_FOR_CHARGE), scheduledBy)
```

File: src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt (hunk 6 above)
Comment (vonElfvin) [reply src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt:15]: Well done!

File: src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+                ScheduleChargeCommand(
+                    memberId = memberId,
+                    scheduledBy = scheduledBy
+                )
+            )
+            if (transition.outcome != SchedulerStateTransitionOutcome.ALLOWED_TRANSITION) {
+                logger.error(transition.message)
+            }
+        }
+
```
Comment (vonElfvin) [inline src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt:42]:
```suggestion
```

File: src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    companion object {
+        private val logger =
+            LoggerFactory.getLogger(ScheduleChargesServiceImpl::class.java)
+    }
```
Comment (vonElfvin) [inline src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt:48]: I personally have no idea how to properly import a logger, I just do this at the top. How come we use a companion object here?

File: src/main/kotlin/com/hedvig/accountservice/schedulers/web/ScheduleController.kt [hunk 10]
```diff
@@ -203,18 +200,16 @@ class ScheduleController @Autowired constructor(
...
-      )
-      if (transition.outcome != SchedulerStateTransitionOutcome.ALLOWED_TRANSITION) {
-        logger.error { transition.message }
-      }
-    }
+    scheduleChargesService.scheduleChargesForAll(scheduledBy)
+    return ResponseEntity.noContent().build()
+  }
+
+  @PostMapping("paymentProvider/{paymentProvider}/charges")
```
Comment (vonElfvin) [inline src/main/kotlin/com/hedvig/accountservice/schedulers/web/ScheduleController.kt:207]: I like it to be structured like this. Do as you please though!
```suggestion
  @PostMapping("charges/provider/{paymentProvider}")
```

File: src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt (hunk 9 above)
Comment (FredrikAreschoug) [reply src/main/kotlin/com/hedvig/accountservice/schedulers/service/ScheduleChargesServiceImpl.kt:48]: Maybe it's an old java habit but if something can be static it should be. My guess is that kotlin figures this out in that it's a static field in compilation ether way. Guessing old habits die hard

**Pull request merged**